activestorage/app/models/active_storage/preview.rb

```diff
@@ -50,7 +50,9 @@ def initialize(blob, variation_or_variation_key)
   # Processing a preview generates an image from its blob and attaches the preview image to the blob. Because the preview
   # image is stored with the blob, it is only generated once.
   def processed
-    process unless processed?
+    blob.with_lock do
+      process unless processed?
+    end
```

Comment:
Each time the url of a variant is called `processed` gets called.
For example in the [API docs](https://edgeapi.rubyonrails.org/classes/ActiveStorage/Blob/Representable.html#method-i-variant):
```ruby
avatar.variant(resize_to_limit: [100, 100]).processed.url
```

I don't think we want to lock the blob everytime we render a URL when it has already been processed.
```suggestion
    unless processed?
      blob.with_lock do
        process
      end  
    end
```

Reply:
Processing video's can also take some time, which could result in transaction timeouts.
Maybe we should set a timestamp on the blob instead to mark it as processing?

Reply:
Thanks for the suggestion. I tried it, but it still results in an orphaned blob.

We had considered this risk of locking on each check, but it's actually already the case that someone calling `processed` in a view could running the risk of doing all this downloading+work in a web worker so in the event they wound up waiting on a lock by a queue worker, it'd still be less than or equal to the time they'd spend (redundantly) processing the same call to `processed`. And once `processed?` is true, it shouldn't be a particularly long-lived lock.

Given that the current state is a pretty significant data integrity issue, the performance cost of the lock seems preferable. I think if someone has the  time & energy to do so, the "better" solution would be to implement either of the other solutions I mentioned above